src/kernels/MIOpenFractionalMaxPool.cpp

```diff
@@ -0,0 +1,345 @@
+    }
+}
+
+template <typename T, typename Ti>
+__device__ void fractionalMaxPool2dForward(const T* input,
+                                           T* output,
+                                           Ti* indices,
+                                           T* random_sample,
+                                           int64_t KH,
+                                           int64_t KW,
+                                           tensor_view_t<4> input_tv,
+                                           tensor_view_t<4> output_tv,
+                                           tensor_view_t<4> indices_tv,
+                                           tensor_view_t<3> random_sample_tv)
+{
+    uint64_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    tensor_layout_t<4> layout(output_tv, gid);
+    int64_t n, c, oh, ow;
+    n  = layout.layout[0];
+    c  = layout.layout[1];
+    oh = layout.layout[2];
+    ow = layout.layout[3];
+
+    if(n >= output_tv.size[0])
+        return;
+
+    int64_t pool_h = get_interval(
+        CVT_FLOAT2ACCUM(random_sample[random_sample_tv.get_tensor_view_idx({n, c, 0})]),
+        oh,
+        input_tv.size[2],
+        output_tv.size[2],
+        KH);
+    int64_t pool_w = get_interval(
+        CVT_FLOAT2ACCUM(random_sample[random_sample_tv.get_tensor_view_idx({n, c, 1})]),
+        ow,
+        input_tv.size[3],
+        output_tv.size[3],
+        KW);
+
+    FLOAT_ACCUM m = log(0);
+
+    int64_t h_end = pool_h + KH;
+    int64_t w_end = pool_w + KW;
+
+    if(!indices)
+    {
+        for(int64_t h = pool_h; h < h_end; ++h)
+        {
+            for(int64_t w = pool_w; w < w_end; ++w)
+            {
+                FLOAT_ACCUM val =
+                    CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({n, c, h, w})]);
+                if(val > m || isnan(val))
+                {
+                    m = val;
+                }
+            }
+        }
+        output[output_tv.get_tensor_view_idx(layout)] = CVT_ACCUM2FLOAT(m);
+    }
+    else
+    {
+        int64_t mi = pool_h * input_tv.size[3] + pool_w;
+        for(int64_t h = pool_h; h < h_end; ++h)
+        {
+            for(int64_t w = pool_w; w < w_end; ++w)
+            {
+                FLOAT_ACCUM val =
+                    CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({n, c, h, w})]);
+                if(val > m || isnan(val))
+                {
+                    m  = val;
+                    mi = h * input_tv.size[3] + w;
+                }
+            }
+        }
+
+        output[output_tv.get_tensor_view_idx(layout)]   = CVT_ACCUM2FLOAT(m);
+        indices[indices_tv.get_tensor_view_idx(layout)] = static_cast<Ti>(mi);
+    }
+}
+
+extern "C" __global__ void FractionalMaxPool2dForward(const D_TYPE* input,
+                                                      D_TYPE* output,
+                                                      I_TYPE* indices,
+                                                      D_TYPE* random_sample,
```

Comment:
Please also add const for other functions
```suggestion
                                                      const D_TYPE* random_sample,
```

Reply:
resolved